src/WebJobs.Extensions.DurableTask.CodeGeneration/README.md

```diff
@@ -0,0 +1,191 @@
+)
+{
+    var (num1, num2) = context.GetInput<(int, int)>();
+
+    return Task.FromResult(num1 + num2);
+}
+
+[FunctionName("Multiply")]
+public async Task<int> Multiply(
+    [OrchestrationTrigger] IDurableOrchestrationContext context)
+{
+    var (num1, num2) = context.GetInput<(int, int)>();
+
+    var result = 0;
+
+    for (var i = 0; i < num2; i++)
+    {
+        // Depends on Add
+        result += await context.CallActivityAsync<int>("Add", (result, num1));
+    }
+
+    return result;
+}
+```
+
+In another example, the test for the ```Multiply``` Orchestration indirectly depends on the ```Add``` Activity.
+
+```csharp
+[Fact]
+public async Task MultiplyTest()
+{
+    // Arrange
+    var num1 = 5;
+    var num2 = 10;
+    var answer = num1 * num2;
+
+    var mockOrchestrationContext = new Mock<IDurableOrchestrationContext>(MockBehavior.Strict);
+    mockOrchestrationContext.Setup(s => s.GetInput<(int, int)>()).Returns((5, 10));
+    // Depends on Add
+    mockOrchestrationContext.Setup(s => s.CallActivityAsync<int>("Add", new Tuple<int, int>(num1, num2))).ReturnsAsync(answer);
+    var context = mockOrchestrationContext.Object;
+
+    var calculator = new Calculator();
+
+    // Act
+    var result = await calculator.Multiply(context);
+
+    // Assert
+    Assert.Equal(answer, result);
+    mockOrchestrationContext.Verify(c => c.GetInput<(int, int)>(), Times.Once);
+    mockOrchestrationContext.Verify(c => c.CallActivityAsync<int>("Add", new Tuple<int, int>(num1, num2)), Times.Once);
+}
+```
+
+## The Solution
+
+Using the recently released resource Roslyn-based [Source Generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) we can automatically generate code to abstract the orchestration/activity call. 
+
+The function name, duable function type (orchestration/activity), parameters (names/types) and return type call can all be discovered by statically analyzing the code. 
+
+An ```IGeneratedDurableOrchestrationCaller``` is generated for representing all Orchestration function calls in the project.
+
+An ```IGeneratedDurableActivityCaller``` is generated for representing all Activity function calls in the project.
+
+An ```IGeneratedDurableOrchestrationStarter``` is generated for representing all Orchestration function starts in the project.
+
+An ```IGeneratedDurableOrchestrationContext``` interface is generated representing an interface for wrapping all operations against previous usage of ```IDurableOrchestrationContext``` and supporting calling Orchestration/Activity functions in the project - exposing an ```IGeneratedDurableOrchestrationCaller``` / ```IGeneratedDurableActivityCaller```.
+
+An ```IGeneratedDurableClient``` interface is generated representing an interface for wrapping all operations against previous usage of ```IDurableClient``` and supporting starting Orchestration functions in the project - exposing ```IGeneratedDurableOrchestrationStarter```.
```

Comment:
Where can I find some examples of this code-generated files?

Reply:
1. There's some examples I put together below for some of the generated code. 
2. If you open the ```WebJobs.Extensions.DurableTask.CodeGen.Example``` project you can see the generated code for those.

I setup the project with the following property in it's csproj:
```xml
<CompilerGeneratedFilesOutputPath>$(BaseIntermediateOutputPath)Generated</CompilerGeneratedFilesOutputPath>
```

You can look in ```WebJobs.Extensions.DurableTask.CodeGen.Example\obj\Generated\WebJobs.Extensions.DurableTask.CodeGen.SourceGenerator``` through an editor of your choice.

Or you can look in the analyzer section in your Visual Studio instance.

![image](https://user-images.githubusercontent.com/7879048/116918211-76541280-ac04-11eb-9f9a-f7cbeab2d38d.png)

I can paste more explicit examples, for all various interfaces/implementations, in the doc if you want.